Add multi-store support

km1tch opened the pull request:
- Store configurations in separate YAML files (store1.yml, store2.yml, etc.)
- Add store-specific API endpoints
- Add price editor UI for modifying box prices
- Update documentation for multi-store functionality
- Enable per-store editability through configuration

km1tch committed:
Add multi-store support

- Store configurations in separate YAML files (store1.yml, store2.yml, etc.)
- Add store-specific API endpoints
- Add price editor UI for modifying box prices
- Update documentation for multi-store functionality
- Enable per-store editability through configuration

navigator9951 commented:
I like the ideas of what you're doing, but I have some suggestions/changes I would prefer before I merge this. Not hard requests, but more RFCs on whether they are reasonable.
- Hide the price editor button when the store cannot be edited
- In the editor: If the entered value is NaN or invalid, bring attention to the cell (maybe change background color)
- Clicking up and down arrows on editor prices sets it to NaN
- Make it so that the no store number defaults to store 1 on page fetch instead of returning a 404
- If there are multiple stores, it may be nicer to move the yaml files to their own directory
- The distinction between `NormalBox` and `CustomBox` doesn't actually need to exist, I just thought it looks nice/readable for the way I stored the data. YAML should be much easier to read so it probably doesn't need that distinction.
- Maybe split out the js for `price_editor.html` to make it similar to `index.html`? If that happens I wouldn't mind creating `index` and `price_editor` directorys for their respective pages

What is alternative depths field for? If it's for cutting down, what the difference between that and the cut down option already built in?
Is there any reasonable way to use `html_content = html_content.replace("Box Price Editor - Store 2", f"Box Price Editor - Store {store_id}")` mechanism to put box info into the intial page load? Is that change in load speed even noticible or impactful?
What is the purpose of the `supplier` and `model` fields?
I'm personally against adding an unauthenticated portal to change the box prices. I think it should require any form of authentication by default. Is there a reason that page doesn't let you change box location (I'm assuming in store/shelf location) and add or remove boxes?

Thoughts? I just want to make sure I'm on the same page as you for some of the stuff you want to do.

km1tch commented:
-i'll see if i can address the 1st 3 items.
-i was trying to make as few changes to the file/folder structure you had as possible.  i agree. YAML (and other files like store floorplans) should be broken out
-I kept NormalBox/Customerbox to keep your format working as-is.  I agree, I don't see the need for the distinction either.  -Our box vendor always labels boxes LxWxD and D is always the open end.
-Again I was trying to minimize changes, and again, I agree.  Probably time for some refactoring for proper separation of concerns.

Alt Depths are the depths that *some* of our boxes are pre-scored from the manufacturer (along the Z-axis).  Most of our employees can't/won't cut down a box to other depths, but these depths are easy.  My plan is to make TWO cut-down options, one for the factory scores and one if you're willing to score it yourself to an arbitrary depth (ie the cut-down you have now in your tool)

The one optimization I did was to load all the box data in one GET request.  Once I did that, I didn't notice any loading time so I didn't optimize further.  We could, it there's a problem.

Supplier is where the boxes came from.  We get ours from America's Box Choice (ABC).  Every box there has a model#.  Some of them are flagged as "multi-depth" (alt-depths above)  See:  https://www.myboxchoice.com/index.php?main_page=index&cPath=1_115 for some examples.  My thought was to generalize this to work for other suppliers (who likely also have some SKU for each box) and potentially make this crowdsourceable, so that once someone had scraped/entered boxed for ABC/other vendors that all that would need to be changed was price/location.  And again, one of my design principles was to keep backwards compat with the formats you used in your version, so supplier/model is optional and (if you look at the next PR) prices[4] still works even though we need to have more itemized-pricing

My thoughts on auth mirror yours, but we're at the phase of the project that I need to ask the store managers to enter the prices.  Rather than have them enter them in an XLS and then have to import them, I figured I'd have them enter them in the tool, and then I'd "turn off" the edit prop in the YAML file.  My short-term plan is just to have this tool have low-friction (no auth) for employees using it in R/O mode.  Still mulling this over, TBH but any layer of friction makes it harder to get adoption by employees.  For now I just "kicked that can down the road"

Re: Location.  I've polled a few managers at our stores and they are divided on whether they WANT to track inventory.  Very do-able, but they don't track inventory now.  This would be a MAJOR operational change for them.  And it WOULD really require auth for the tool.  For now, I'm leaning towards not tracking quantities, but we'll see if that thought changes.  What I DO want to do is to have each store (optionally) upload a floorplan, and to have the location store the coords on the floorplan for where the box can be found.  Then render the "location" as an X-marked floorplan.  Should help with finding some of the less commonly used boxes.  For now it's more of a placeholder.  The plan will be to have it have it's own edit feature where you can mark an X on a floorplan to save.

Happy to get on a call to discuss any of this if you'd like.  And if our visions for this diverge too much, just let me know.  Not trying to strongarm you into any changes to your app, just inspired to make it solve our use case.  Next big up is more of a "novice view" which helps an employee pick a box--vs what now is more of an "I know what I'm doing" view  :)

km1tch committed:
Fix UI issues and improve price editor usability

- Hide price editor button when store cannot be edited
- Fix invalid input handling in price editor
- Remove up/down arrows from number inputs to prevent NaN values
- Improve validation of price inputs

navigator9951 commented:
LGTM for your changes. 
> LxWxD and D is always the open end

Makes sense to me, our store orders all boxes in order of dimension length without caring about which dimension opens. I have no attachments to preserving the system just to match our sheets.

> Auth

Seems reasonable to me. I wouldn't expect any malicious things to occour, so we can deal with that later.

> Inventory

Our store doesn't track inventory and tbh I don't want to either. While having a floor plan does sound neat, I'm not sure it would be too useful since the final stage of the site is to print out packing info on a thermal label. It probably wouldn't print/scale well. I think a text field for location should be plenty since you can just describe locations or just name in-store locations various letters of the alphabet.

Tbh as long as the functionality I need doesn't degrade, I don't mind anything you want to do. I do read through all changes you make since I want to know what you're trying to do and to understand what I'm running on my machines.

My suggestion to people that don't want to read a bunch of things, the "novice view" if you will, is just to uncheck the "show possible boxes" option, only check the packing level they need, and print the top result. That way only boxes that are guarenteed to fit as needed will show up.
